pyuvdata/uvbeam.py

```diff
@@ -410,6 +412,74 @@ def set_phased_array(self):
         self._gain_array.required = True
         self._coupling_matrix.required = True
 
+    def az_za_to_healpix(self, nside=None):
+        """
+        Convert beam in az_za coordinates to healpix coordinates. Pixelization is done
+        using healpy's ang2pix method which is presumably a nearest neighbor approach.
+
+        Args:
+            nside: The nside to use for the Healpix map. If not specified, use
+            the nside that gives the closest resolution that is higher than the
+            input resolution.
+        """
+        import healpy as hp
+        if self.pixel_coordinate_system != 'az_za':
+            raise ValueError('pixel_coordinate_system must be "az_za"')
+        if self.beam_type != 'power':
+            raise ValueError('healpix conversion not yet defined for efield')
+
+        phi_vals, theta_vals = np.meshgrid(self.axis1_array, self.axis2_array)
+
+        if nside is None:
+            min_res = np.min(np.array([np.diff(self.axis1_array)[0], np.diff(self.axis2_array)[0]]))
+            nside_min_res = np.sqrt(3 / np.pi) * np.radians(60.) / min_res
+            nside = int(2**np.ceil(np.log2(nside_min_res)))
+            assert(hp.pixelfunc.nside2resol(nside) < min_res)
+
+        npix = hp.nside2npix(nside)
+
+        az_za_data_shape = self.data_array.shape
+        new_shape = list(az_za_data_shape[0:4])
+        new_shape.append(self.Naxes1 * self.Naxes2)
+        new_shape = tuple(new_shape)
+        az_za_data = copy.deepcopy(self.data_array).reshape(new_shape)
+        phi_vals = phi_vals.flatten()
+        theta_vals = theta_vals.flatten()
+
+        self.pixel_coordinate_system = 'healpix'
+        self.nside = nside
+        self.Npixels = npix
+        self.ordering = 'ring'
+        self.set_cs_params()
+
+        healpix_data = np.zeros(self._data_array.expected_shape(self), dtype=np.float)
+        data_pixels = hp.ang2pix(nside, theta_vals, phi_vals)
+        pixels = np.arange(npix)
+        hits = np.zeros(self._data_array.expected_shape(self))
+
+        for index in range(self.Naxes1 * self.Naxes2):
+            healpix_data[:, :, :, :, data_pixels[index]] += az_za_data[:, :, :, :, index]
+            hits[:, :, :, :, data_pixels[index]] += 1
```

Comment:
Hmm, now I'm not sure about this interpolation method. When we were undersampling, each covered healpix pixel would get a hit. But now that we're oversampling, we are missing healpix pixels between az_za pixels. For example, about 15% of the pixels get missed when converting Nic's 150MHz beam. If this is desired, that's fine, but it seems to me like we would want to fill those in with interpolation.

Reply:
yeah, I was worried about that too. Let me take a look at what Ian does in FHD.

Reply:
fwiw, because we require az_za are on a regular grid, you can use some pretty efficient interpolation routines. I have some code here which did this using RectBivariateSpline:
https://github.com/HERA-Team/HERA-Beams/blob/master/NicolasFagnoniBeams/convert_to_healpix.py

Reply:
So I'm worried about a couple of edge cases. I think actually that the regular grid requirement should not be enforced on the object, but only if it's written to a regularly gridded FITS file. So I think those errors should actually be warnings. Also, it's definitely not required to cover the sphere and if it doesn't, the healpix pixels outside the covered area should just be excluded.

Reply:
Hmm.. I'm wrong about the regular grid -- that is baked into the object. I was confused because we had talked about doing something else, but we ultimately decided that was a reasonable requirement. I'm still worried about it not covering the sphere though.